### PR TITLE
PWGHF: Add Ds task and selector

### PR DESCRIPTION
**Pull request by stefanopolitano**
Adding Ds+ -> phi pi -> pi K K task following Dplus example.

### CODE REVIEW

**Review by vkucera — CHANGES_REQUESTED** [review]

**Comment by vkucera** [comment]
Also please create a PR in the validation repository.

**Comment by stefanopolitano** [comment]
Hi @vkucera and @ginnocen! Sorry for the delay: I have been a little busy finalizing the analyses for the next week PWG-HF meeting. I am addressing the comments: I will push the correction asap

**Comment by vkucera** [comment]
Hi @stefanopolitano , any updates?

**Comment by stefanopolitano** [comment]
Hi @vkucera! I finished addressing your comments, but I still have a doubt about this suggestion:

> Change the `struct` name so that `TaskName` is not needed.>

Indeed, it seems that the same naming approach is used in the others PWGHF tasks and selectors. Could you give me an example of the correct way to rename the struct?

**Comment by vkucera** [comment]
> Hi @vkucera! I finished addressing your comments, but I still have a doubt about this suggestion:
> 
> > Change the `struct` name so that `TaskName` is not needed.>
> 
> Indeed, it seems that the same naming approach is used in the others PWGHF tasks and selectors. Could you give me an example of the correct way to rename the struct?

Hi @stefanopolitano , [here](https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/Tasks/taskBPlus.cxx#L45) is an example. In the absence of `TaskName`, the device name is derived from the `struct` name by replacing upper-case letters with hyphens and lower-case letters.

**Comment by stefanopolitano** [comment]
PWGHF: updating Ds task according to Dplus task example and addressing comments on the first version of Ds task and selector.

File: PWGHF/DataModel/HFCandidateSelectionTables.h [hunk 1]
```diff
@@ -90,9 +90,16 @@ namespace hf_selcandidate_dplus
 {
 DECLARE_SOA_COLUMN(IsSelDplusToPiKPi, isSelDplusToPiKPi, int); //!
 } // namespace hf_selcandidate_dplus
-DECLARE_SOA_TABLE(HFSelDplusToPiKPiCandidate, "AOD", "HFSELDPLUSCAND", //!
+DECLARE_SOA_TABLE(HFSelDplusToPiKPiCandidate, "AOD", "HFSELdpcandidateCAND", //!
```
Comment (fcolamar) [inline PWGHF/DataModel/HFCandidateSelectionTables.h]: I think this is an unwanted modification

File: PWGHF/Tasks/taskDs.cxx [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    }
+  }
+  PROCESS_SWITCH(HfTaskDs, processMC, "Process MC", false);
+};
+
+WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
+{
+  WorkflowSpec workflow{};
+  workflow.push_back(adaptAnalysisTask<HfTaskDs>(cfgc));
+  return workflow;
```
Comment (fcolamar) [inline PWGHF/Tasks/taskDs.cxx:220]:
```suggestion
  return WorkflowSpec{adaptAnalysisTask<HfTaskDs>(cfgc)};
```
More compact

**Review by fcolamar — CHANGES_REQUESTED** [review]
Hello Stefano,
some further comments on the taskDs.cxx. After addressing these, the task would be ok to me

File: PWGHF/DataModel/HFCandidateSelectionTables.h (hunk 1 above)
Comment (fcolamar) [reply PWGHF/DataModel/HFCandidateSelectionTables.h]: ->I saw that you just fixed this while I was finishing to review the task

**Comment by fcolamar** [comment]
Hi @stefanopolitano, thanks for the updates, could you also reply to the questions from the taskD0 review where you didn't modify the code? (for the others there's no need since you already took action, and I resolved those comments)

File: PWGHF/TableProducer/HFCandidateCreator3Prong.cxx [hunk 3]
```diff
@@ -242,6 +250,14 @@ struct HFCandidateCreator3ProngExpressions {
         flag = sign * (1 << DecayType::DPlusToPiKPi);
       }
 
+      // Ds± → π± K∓ π±
+      //Printf("Checking Ds± → K± K∓ π±");
+      if (flag == 0) {
```
Comment (vkucera) [inline PWGHF/TableProducer/HFCandidateCreator3Prong.cxx:255]: The `Printf` has to be inside the block to be useful.

File: PWGHF/TableProducer/HFCandidateCreator3Prong.cxx [hunk 4]
```diff
@@ -184,6 +184,14 @@ struct HFCandidateCreator3ProngExpressions {
         flag = sign * (1 << DecayType::DPlusToPiKPi);
       }
 
+      // Ds± → π± K∓ π±
```
Comment (vkucera) [inline PWGHF/TableProducer/HFCandidateCreator3Prong.cxx:187]: The decay is wrong.

File: PWGHF/DataModel/HFSecondaryVertex.h [hunk 5]
```diff
@@ -611,7 +611,7 @@ DECLARE_SOA_COLUMN(FlagMCDecayChanGen, flagMCDecayChanGen, int8_t); //! resonant
 // mapping of decay types
 enum DecayType { DPlusToPiKPi = 0,
                  LcToPKPi,
-                 DsToPiKK,
+                 DsToKKPi,
```
Comment (vkucera) [inline PWGHF/DataModel/HFSecondaryVertex.h:614]: Why?

File: PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    if (trackPion1.pt() < cuts->get(pTBin, "pT Pi") || trackKaon.pt() < cuts->get(pTBin, "pT K") || trackPion2.pt() < cuts->get(pTBin, "pT Pi")) {
+      return false;
+    }
+    // invariant-mass cut
+    if (std::abs(InvMassDsKKPi(candidate) - RecoDecay::getMassPDG(pdg::Code::kDs)) > cuts->get(pTBin, "deltaM")) {
+      return false;
+    }
+    if (std::abs(InvMassDsPiKK(candidate) - RecoDecay::getMassPDG(pdg::Code::kDs)) > cuts->get(pTBin, "deltaM")) {
+      return false;
+    }
```
Comment (vkucera) [inline PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx:76]: This seems wrong. You reject the candidate if it doesn't satisfy both mass hypotheses.

File: PWGHF/DataModel/HFSecondaryVertex.h [hunk 7]
```diff
@@ -425,7 +425,7 @@ auto InvMassGamma(const T& candidate)
 } // namespace hf_cand_casc
 
 DECLARE_SOA_TABLE(HfCandCascBase, "AOD", "HFCANDCASCBASE", //!
-                  // general columns
+                                                           // general columns
```
Comment (vkucera) [inline PWGHF/DataModel/HFSecondaryVertex.h:428]: Revert

File: PWGHF/DataModel/HFCandidateSelectionTables.h [hunk 8]
```diff
@@ -93,6 +93,13 @@ DECLARE_SOA_COLUMN(IsSelDplusToPiKPi, isSelDplusToPiKPi, int); //!
 DECLARE_SOA_TABLE(HFSelDplusToPiKPiCandidate, "AOD", "HFSELDPLUSCAND", //!
                   hf_selcandidate_dplus::IsSelDplusToPiKPi);
 
+namespace hf_selcandidate_ds
+{
+DECLARE_SOA_COLUMN(IsSelDsToKKPi, isSelDsToKKPi, int); //!
+} // namespace hf_selcandidate_ds
```
Comment (vkucera) [inline PWGHF/DataModel/HFCandidateSelectionTables.h:99]: How come that you don't need `isSelDsToPiKK`?

File: PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+      SETBIT(statusDsToKKPi, aod::SelectionStep::RecoSkims);
+
+      auto trackPos1 = candidate.index0_as<aod::BigTracksPID>(); // positive daughter (negative for the antiparticles)
+      auto trackNeg = candidate.index1_as<aod::BigTracksPID>();  // negative daughter (positive for the antiparticles)
+      auto trackPos2 = candidate.index2_as<aod::BigTracksPID>(); // positive daughter (negative for the antiparticles)
+
+      // topological selection
+      if ((!selection(candidate, trackPos1, trackNeg, trackPos2)) || !selection(candidate, trackPos2, trackNeg, trackPos1)) {
+        hfSelDsToKKPiCandidate(statusDsToKKPi);
+        continue;
```
Comment (vkucera) [inline PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx:127]: Why do you reject the candidate if it doesn't satisfy the selection with both track permutations?

File: PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+      // track-level PID selection
+      int pidTrackPos1Pion = selectorPion.getStatusTrackPIDAll(trackPos1);
+      int pidTrackNegKaon = selectorKaon.getStatusTrackPIDAll(trackNeg);
+      int pidTrackPos2Pion = selectorPion.getStatusTrackPIDAll(trackPos2);
+
+      if (pidTrackPos1Pion == TrackSelectorPID::Status::PIDRejected ||
+          pidTrackNegKaon == TrackSelectorPID::Status::PIDRejected ||
+          pidTrackPos2Pion == TrackSelectorPID::Status::PIDRejected) { // exclude D±
+        hfSelDsToKKPiCandidate(statusDsToKKPi);
+        continue;
```
Comment (vkucera) [inline PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx:140]: Why do you want the first and last track to be pions?

File: PWGHF/Tasks/CMakeLists.txt [hunk 11]
```diff
@@ -44,6 +44,11 @@ o2physics_add_dpl_workflow(task-dplus
                     PUBLIC_LINK_LIBRARIES O2::Framework O2Physics::AnalysisCore O2::DetectorsVertexing
                     COMPONENT_NAME Analysis)
 
+o2physics_add_dpl_workflow(task-ds
+                    SOURCES taskDs.cxx
+                    PUBLIC_LINK_LIBRARIES O2::Framework O2Physics::AnalysisCore O2::DetectorsVertexing
+                    COMPONENT_NAME Analysis)
+
```
Comment (vkucera) [inline PWGHF/Tasks/CMakeLists.txt:51]:
```suggestion
```
Do not add unnecessary extra lines.

File: PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+      SETBIT(statusDsToKKPi, aod::SelectionStep::RecoTopol);
+
+      // track-level PID selection
+      int pidTrackPos1Pion = selectorPion.getStatusTrackPIDAll(trackPos1);
+      int pidTrackNegKaon = selectorKaon.getStatusTrackPIDAll(trackNeg);
+      int pidTrackPos2Pion = selectorPion.getStatusTrackPIDAll(trackPos2);
+
+      if (pidTrackPos1Pion == TrackSelectorPID::Status::PIDRejected ||
+          pidTrackNegKaon == TrackSelectorPID::Status::PIDRejected ||
+          pidTrackPos2Pion == TrackSelectorPID::Status::PIDRejected) { // exclude D±
```
Comment (vkucera) [inline PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx:138]: > // exclude D±

Ds?

**Review by vkucera — CHANGES_REQUESTED** [review]

**Comment by fcolamar** [comment]
HI @stefanopolitano, I see that you pushed some updates (though maybe you recreated the whole files, since checking the diff of last commit everything appears as new), can you also reply to the pending conversations on this page? Thanks!

**Comment by stefanopolitano** [comment]
Hi @fcolamar! Yes, sorry for the delay: I will reply to the pending conversations immediately. Thanks to you!

**Comment by stefanopolitano** [comment]
Since the conversations refer to outdated code, it seems I can not answer directly to them so I will reply in this comment. 
For what concerns this question (https://github.com/AliceO2Group/O2Physics/pull/96#discussion_r759422216), yes it works: hf_cuts_ds_tokkpi is called just once in taskDs to define the number of pT bins. Indeed, the information on the cuts is needed in the HFDsToKKPiCandidateSelector. 
For what concerns this question (https://github.com/AliceO2Group/O2Physics/pull/96#discussion_r759431746), I don't have a strong opinion on this.  At the moment, I am keeping them as TH1 for consistency with the taskDplus but there would be no problem in having these as TH2 histograms.

File: PWGHF/DataModel/HFCandidateSelectionTables.h (hunk 8 above)
Comment (vkucera) [reply PWGHF/DataModel/HFCandidateSelectionTables.h]: I don't see `IsSelDsToPiKK` used in the filters.

File: PWGHF/DataModel/HFSecondaryVertex.h (hunk 5 above)
Comment (vkucera) [reply PWGHF/DataModel/HFSecondaryVertex.h:614]: It's not a big deal but I am still wondering what is the reason for this change.

File: PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+        pidDsKKPi = 0; // exclude DsKKPi
+      }
+      // checking PID for Ds to PiKK hypothesis
+      if (pidTrackPos1Pion == TrackSelectorPID::Status::PIDAccepted &&
+          pidTrackNegKaon == TrackSelectorPID::Status::PIDAccepted &&
+          pidTrackPos2Kaon == TrackSelectorPID::Status::PIDAccepted) {
+        pidDsPiKK = 1; // accept DsPiKK
+      } else if (pidTrackPos1Pion == TrackSelectorPID::Status::PIDRejected ||
+                 pidTrackPos2Kaon == TrackSelectorPID::Status::PIDRejected) {
+        pidDsPiKK = 0; // exclude DsKKPi
```
Comment (vkucera) [inline PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx:160]:
```suggestion
        pidDsPiKK = 0; // exclude DsPiKK
```

File: PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+        continue;
+      }
+      SETBIT(statusDsToKKPi, aod::SelectionStep::RecoPID);
+
+      if ((pidDsKKPi == 1 || pidDsKKPi == -1) && selection(candidate, trackPos1, trackNeg, trackPos2)) {
+        statusDsToKKPi = 1; // Ds to KKPi hypotesis matched
+      }
+      if ((pidDsPiKK == 1 || pidDsPiKK == -1) && selection(candidate, trackPos2, trackNeg, trackPos1)) {
+        statusDsToPiKK = 1; // Ds to PiKK hypotesis matched
+      }
```
Comment (vkucera) [inline PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx:173]: Why do you call `selection` again? You already did on line 123.

**Comment by vkucera** [comment]
> Since the conversations refer to outdated code, it seems I can not answer directly to them so I will reply in this comment. For what concerns this question ([#96 (comment)](https://github.com/AliceO2Group/O2Physics/pull/96#discussion_r759422216)), yes it works: hf_cuts_ds_tokkpi is called just once in taskDs to define the number of pT bins. Indeed, the information on the cuts is needed in the HFDsToKKPiCandidateSelector. For what concerns this question ([#96 (comment)](https://github.com/AliceO2Group/O2Physics/pull/96#discussion_r759431746)), I don't have a strong opinion on this. At the moment, I am keeping them as TH1 for consistency with the taskDplus but there would be no problem in having these as TH2 histograms.

Hi @stefanopolitano , thanks for the updates. Next time, please don't overwrite existing commits that have been already reviewed. It makes it more difficult to compare the latest version with the reviewed one. Also please reply directly to a comment when you address it.
Have a look at the new and remaining comments.

**Pull request closed without merging**